lib/mapObjects/CGHeroInstance.cpp

```diff
@@ -129,14 +129,9 @@ int3 CGHeroInstance::getPosition(bool h3m) const //h3m=true - returns position o
 	}
 }
 
-bool CGHeroInstance::canFly() const
+const Bonus * CGHeroInstance::getBonusAtTurn(const Bonus::BonusType &type, const int &turn, const TBonusSubtype &subType) const
 {
-	return hasBonusOfType(Bonus::FLYING_MOVEMENT);
-}
-
-bool CGHeroInstance::canWalkOnSea() const
-{
-	return hasBonusOfType(Bonus::WATER_WALKING);
+	return getBonus(Selector::type(type).And(Selector::days(turn)).And(Selector::subtype(subType)));
```

Comment:
I also don`t like the fact that cachingStr is used outside but this is how it works now.